### ISSUE TITLE
Non GTK4 apps blurry on GNOME 46.3.1 overview

### ISSUE THREAD

**Issue by amadeusp** [flexagoon/rounded-window-corners#36]
The contents of non GTK4 application windows are blurry and fuzzy when displayed in GNOME 46.3.1's overview.

# Steps
1. Install the Rounded Window Corners Reborn extension in version 3.
2. Open some GTK4 and some non GTK4 applications and press the super key.
3. The window contents of non GTK4 apps are blurry and fuzzy.

# Expected behavior
All types of applications should be displayed with the same quality.

# My system
* Manjaro Linux x86_64
* Linux 6.10.0-1-MANJARO
* GNOME 46.3.1
* Mutter (Wayland)
* Fractional scaling is set to 200%
* My display is 32" and 4k

# Screen grabs of non GTK4 apps
![350312467-2674111e-5ffb-4b0b-a6aa-d9df152c9a39](https://github.com/user-attachments/assets/74376f01-1466-4cfe-a29c-db07dcf9dd79)
![350312497-87688ca8-431f-4853-bc53-2503a89dc170](https://github.com/user-attachments/assets/3cae1aa9-922c-46ac-a30d-9c027f7858b7)

**Comment by cyrneko** [comment]
can reproduce! I am gonna try to test a bunch of applications to verify where this behavior does and doesn't happen, to potentially further narrow down what kinds of windows are affected.

**Comment by cyrneko** [comment]
> can reproduce! I am gonna try to test a bunch of applications to verify where this behavior does and doesn't happen, to potentially further narrow down what kinds of windows are affected.

well...after testing pretty much every scenario I can think of, from different versions of GTK to Wayland-Native and XWayland, games and browsers, it really does seem to just be GTK4 apps that are unaffected for...some reason

**Comment by flexagoon** [comment]
@cyrneko gtk4 apps are probably unaffected just because "skip Libadwaita applications" is enabled in extension settings

**Comment by amadeusp** [comment]
> @flexagoon Can confirm on my machine. Thanks for your work on the project!
> 
> ![Screenshot of libadwaita application (settings) looking blurry. "Skip LibAdwaita applications" is toggled off.](https://private-user-images.githubusercontent.com/51167014/353109053-2032639c-b838-4b0e-9924-b05ab48e3cd7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjIyNjQxMTQsIm5iZiI6MTcyMjI2MzgxNCwicGF0aCI6Ii81MTE2NzAxNC8zNTMxMDkwNTMtMjAzMjYzOWMtYjgzOC00YjBlLTk5MjQtYjA1YWI0OGUzY2Q3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzI5VDE0MzY1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk1ZTAxOWY3MmVhMTBiNGI5NTY3N2M4NzI1N2RmZWE5MDFhMTJmZDNhMTcxZDdmNjAxZTY0ZmI2MThhN2Y5Y2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.B2-9s5GaiCXII_a88oybjTG2fn8QuaKVt_wU_zz5sB8)
> 
> ![Screenshot of libadwaita application (settings) looking sharp and clean. "Skip LibAdwaita applications" is toggled on.](https://private-user-images.githubusercontent.com/51167014/353109217-a8f9f601-da7f-40c5-b84e-56cb081abe0a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjIyNjQxMTQsIm5iZiI6MTcyMjI2MzgxNCwicGF0aCI6Ii81MTE2NzAxNC8zNTMxMDkyMTctYThmOWY2MDEtZGE3Zi00MGM1LWI4NGUtNTZjYjA4MWFiZTBhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzI5VDE0MzY1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU2Nzg4YmU4NjkzNjU5ODE5YzhmZWFiN2Y2MTY1Y2FmM2UxZDExM2JhNDQ3M2RjZmExMjgxM2QwODY4NWQ5MzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.LB43_wDmOwqe-stAOnC7xcCiytKP9WAvH1ozgRhqNZU)
> 
> Non libadwaita apps (tested with thunderbird) still look blurry with "Skip LibAdwaita applications" toggled on.

I think [flexagoon](https://github.com/flexagoon) meant that the only reason GTK4 apps were not blurry in our tests was because they were turned off via the "Skip LibAdwaita applications" setting.

**Comment by Impasse52** [comment]
I'm experiencing the same behavior even without any scaling at all, on Gnome 46, Fedora 40.

**Comment by BlackRabbit22** [comment]
I'm also experiencing this without any scaling. It affects both GTk3 and GTK4 applications.

**Comment by konstantintutsch** [comment]
I think this issue seems to be related to how large the windows are calculated to be scaled in the overview.

I've noticed, at least on my end, that the issue only occurs when two (or more) applications are opened on two (or more) different workspaces.

Here's a video showcasing this:

https://github.com/user-attachments/assets/1b65f044-dae8-4a03-9abe-f412eb11aed8

The extension might be calculating and rendering the size of each window in the overview across all workspaces, not separate for each workspace and it's windows.

**Comment by flexagoon** [comment]
@konstantintutsch the issue is caused by commit https://github.com/flexagoon/rounded-window-corners/commit/1c0e89cbfd42703f95b42aca3f91cc0e8fd237b4 and is explained in the description of that commit:

> **fix: display windows correctly in overview**
> 
> Previously, when opening the overview, the corner rounding was removed
from windows themselves and applied to window previews.
> 
> This prevented the windows from being slightly blurry, but after a
Mutter update, the code that was handling this broke, and window preview
sizes were calculated incorrectly.
> 
> This commit removes this behavior. It means that windows in overview
will be a bit pixelated, but it isn't very noticeable, and certainly
better than the windows being half transparent. Maybe the blurriness
issue will be easier to fix in the future, after I refactor the
codebase.

**Comment by konstantintutsch** [comment]
@flexagoon Ah okay, sorry to have bothered you. Thought this issue was still relevant 😅

**Comment by flexagoon** [comment]
@konstantintutsch no, you didn't bother me and the issue is indeed relevant, I just currently don't know how to fix it unfortunately

**Comment by aunetx** [comment]
This happens because of a bug in upstream GNOME (quite easy to reproduce, steps described in the issue I opened): https://gitlab.gnome.org/GNOME/gnome-shell/-/issues/7903

I've not yet understood how to hack beyond it, but if I find how I will tell you!

**Comment by Neo-29** [comment]
> This happens because of a bug in upstream GNOME (quite easy to reproduce, steps described in the issue I opened): https://gitlab.gnome.org/GNOME/gnome-shell/-/issues/7903
> 
> I've not yet understood how to hack beyond it, but if I find how I will tell you!

Here's my comment with the link to this post, if that might help...

**Comment by flexagoon** [comment]
I'm locking this since it's an upstream issue and it's up to Gnome to fix this

@aunetx please ping me somewhere else if you find a way to work around this
